interfaces/apparmor/backend.go

```diff
@@ -152,6 +152,19 @@ func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions,
 			logger.Noticef("cannot create host snap-confine apparmor configuration: %s", err)
 		}
 	}
+	// core on core devices is also special, the apparmor cache gets
+	// confused too easy, especially at rollbacks, so we delete the cache.
+	// See LP:#1460152 and
+	// https://forum.snapcraft.io/t/core-snap-revert-issues-on-core-devices/
+	if snapName == "core" && !release.OnClassic {
+		if l, err := filepath.Glob(dirs.SystemApparmorCacheDir + "/*"); err == nil {
+			for _, p := range l {
```

Comment:
Nitpick: while there shouldn't be any non-files in /etc/apparmor.d/cache, I wondered if it would make sense to verify if [IsRegular](https://golang.org/pkg/os/#FileMode.IsRegular). You are checking the return value of os.Remove(), so this isn't strictly necessary, and looking at the [implementation](https://golang.org/src/os/file_unix.go?s=7909:7939#L258), it just calls unlink followed by rmdir, so there isn't a security risk with what you are doing. Not a blocker; just mentioning for your consideration.